lib/alexa_request_verifier.ex

```diff
@@ -1,283 +1,339 @@
         if is_nil(cert) do
           cert = fetch_cert(cert_url)
-        case validate_cert(cert) do
-
-        {:ok, _} ->
-            Logger.debug("alexa_request_verifier: setting cache for #{cert_url}")
-            ConCache.put(:cert_signature_cache, cert_url, cert)
-
-            Conn.put_private(conn,:signing_cert, cert)
-              
-        {:error, reason} ->
-            Logger.debug("alexa_request_verifier: got error -  #{reason}")
-            Conn.put_private(conn, :alexa_verify_error, reason )
-              
+
+          case validate_cert(cert) do
+            {:ok, _} ->
+              Logger.debug(
+                "alexa_request_verifier: caching url",
+                cert_url: cert_url
+              )
+
+              CertCache.store(cert_url, cert)
+              Conn.put_private(conn, :signing_cert, cert)
+
+            {:error, reason} ->
+              Logger.debug(
+                "alexa_request_verifier validation error",
+                validation_error: reason
+              )
+
+              Conn.put_private(conn, :alexa_verify_error, reason)
           end
         else
-          Conn.put_private(conn,:signing_cert, cert)
+          Conn.put_private(conn, :signing_cert, cert)
         end
     end
-end
+  end
 
-  
+  defp fetch_cert(url) do
+    if is_correct_alexa_url?(url) do
+      {:ok, resp} =
+        :httpc.request(
+          :get,
+          {String.to_charlist(url), []},
+          [],
+          body_format: :binary
+        )
 
-@doc """
-  takes a string and confirms URL is in scheme https, has a s3.amazonaws.com host, is port 443 and has a path starting with /echo.api/
+      {_, _headers, certificate_chain_bin} = resp
+      cert_chain = :public_key.pem_decode(certificate_chain_bin)
 
-"""
+      Enum.map(
+        cert_chain,
+        fn {_, bin, _} -> bin end
+      )
+    else
+      {:error, "invalid sig chain url"}
+    end
+  end
 
+  @doc """
+  Determines whether a request URL represents a valid
+  Alexa request.
+  """
+  @spec is_correct_alexa_url?(url :: String.t() | URI.t()) :: boolean
   def is_correct_alexa_url?(url) when is_binary(url) do
-      is_correct_alexa_url?(URI.parse(url))
-
+    is_correct_alexa_url?(URI.parse(url))
   end
-  
 
   def is_correct_alexa_url?(url) when is_nil(url) do
-      false
+    false
   end
 
-
-  def is_correct_alexa_url?(%URI{port: 443, host: "s3.amazonaws.com", scheme: "https", path: "/echo.api/" <> _extra}) do
-      true  
+  def is_correct_alexa_url?(%URI{
+        port: 443,
+        host: "s3.amazonaws.com",
+        scheme: "https",
+        path: "/echo.api/" <> _extra
+      }) do
+    true
   end
 
   def is_correct_alexa_url?(_everything_else) do
-      false  
+    false
   end
 
+  defp validate_cert(err = {:error, _reason}) do
+    err
+  end
 
-  def fetch_cert(url) do
-    if(!is_correct_alexa_url?(url)) do
-      {:error, "invalid sig chain url"}
+  defp validate_cert(cert) do
```

Comment:
This is really where the caching should be done. You don't want to verify the cert chain on every request. Once you have verified the cert chain, cache that result, so you can just do a lookup on the cert in subsequent requests to check that it is valid.